### PR TITLE
[lit-html] add Lit 2 directive syntax to Lit 1

### PR DESCRIPTION
**Pull request by e111077**
Add forwards compatibility for Lit 1 directives. This will allow users to write with the Lit 2 directive class syntax and also adds some files so that the we can migrate users to the new import targets and then switch them to Lit 2 rather than those changes being atomic.

Notable changes:
* `directive` can now take a `DirectiveClass` as an argument in `lit-html/directives`
* The API of the lit 2 `[a-zA-Z]+Part`s have added to `lit-html/directives`
* creation of root level:
  * `async-directive.ts`
  * `directive.ts`
* creation of `ChildPart`

Note: this seems to have broken lit-analyzer 1.1.9 internally, but i could not reproduce those errors externally. This seems to work just fine in lit-analyzer 1.2.0.

Additionally, usage of `ChildPart` in a directive may cause lit-analyzer to not detect that a directive is a directive, but this should not cause any failures because `NodePart` (Lit 1 equivalent) is never bound to anything that needs type checking.

Related #1653

### CODE REVIEW

File: src/directive.ts
Comment (kevinpschaaf) [inline src/directive.ts:70]: Unclear why this is exported? It's not v2 api...

File: src/async-directive.ts
Comment (kevinpschaaf) [inline src/async-directive.ts:76]: Docs seem wrong here... they're never called right?

File: src/directive.ts
Comment (kevinpschaaf) [inline src/directive.ts:87]: For a very modest code size increase, we could go and put `options` on each legacy part type (the `DefaultTemplateProcessor` that creates all the parts has it) via an optional constructor argument on each Part class, so that the v2 parts can actually return something here.

It still won't guarantee it's there for people who created custom template processors, but seems like it would be an improvement (and they can't migrate that code to v2 anyway)?

Or maybe we just don't care because they couldn't have relied on options in v1, so it's not a migration concern?

@justinfagnani thoughts?

**Review by kevinpschaaf — COMMENTED** [review]
Just a few minor questions

File: src/directive.ts
Comment (e111077) [reply src/directive.ts:70]: You're right. This was a copy paste will clean up

File: src/async-directive.ts
Comment (e111077) [reply src/async-directive.ts:76]: @rictic

**Comment by kevinpschaaf** [comment]
We also need to add forward-compat for at least `directive-helpers.js`:`isTemplateResult` (needed for open-wc/storybook to migrate to hybrid support 1.x and 2.x).

We should also consider adding these to the `directive-helpers.js`:
* `isTemplateResult`
* `insertPart`
* `setChildPartValue`
* `setCommittedValue`
* `getCommittedValue`
* `removePart`
* `clearPart`

Seems like they would be required to migrate more advanced directives.

File: src/directive-helpers.ts
Comment (bicknellr) [inline src/directive-helpers.ts:32]: Just a heads up, we shouldn't need `getRenderedNodes` anymore now that #1642 is merged.

File: src/async-directive.ts
Comment (kevinpschaaf) [reply src/async-directive.ts:76]: They won't get called in V1 (we should just document that fact in the API docs). The idea is this is ok because a V1 directive couldn't have relied on the callback being called.

The only reason to put them here on the interface is so that users can go ahead and write code that _will_ have the callbacks run in V2.

Comment (rictic) [reply src/async-directive.ts:76]: Yeah, sorry, this doc is out of date. I think for a little while I have a halfway implementation here, but it diverged enough from the behavior of the v2 version that it seemed like it might do more harm than good.

+1 to having clear docs that these are just for interface compatibility, that lit-html v1 will not call them

File: src/directive.ts
Comment (rictic) [reply src/directive.ts:75]: IMO since this is for allowing existing directives to migrate to the new API surface, we don't need to support novel features, just translation

Comment (kevinpschaaf) [reply src/directive.ts:75]: Yeah, makes sense.

File: src/directive.ts
Comment (justinfagnani) [inline src/directive.ts:35]: looks like this needs formatting

File: src/directive.ts
Comment (justinfagnani) [reply src/directive.ts:75]: I agree w/ @rictic here

File: src/async-directive.ts
Comment (kevinpschaaf) [inline src/async-directive.ts:77]:
```suggestion
  /**
   * User callback for implementing logic to release any
   * resources/subscriptions that may have been retained by this directive.
   * Since directives may also be re-connected, `reconnected` should also be
   * implemented to restore the working state of the directive prior to the next
   * render.
   *
   * NOTE: In lit-html 1.x, the `disconnected` and `reconnected` callbacks WILL NOT BE
   * CALLED. The interface is provided here for forward-compatible directive authoring only.
   */
```

File: src/directive.ts
Comment (justinfagnani) [inline src/directive.ts:134]: Don't use constructor parameter properties - declare this as a regular property.

File: src/async-directive.ts
Comment (kevinpschaaf) [inline src/async-directive.ts:82]:
```suggestion
  /**
   * User callback to restore the working state of the directive prior to the next
   * render. This should generally re-do the work that was undone in `disconnected`.
   *
   * NOTE: In lit-html 1.x, the `disconnected` and `reconnected` callbacks WILL NOT BE
   * CALLED. The interface is provided here for forward-compatible directive authoring only.
   */
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  protected reconnected() {
```

**Review by kevinpschaaf — DISMISSED** [review]
LGTM, pending CI green(formatting issues?).

Added some suggestions for the API docs for the callbacks (I realize the lit-next ones only had one API doc for both callbacks; we'll have to fix it there also.

File: src/directive.ts
Comment (justinfagnani) [inline src/directive.ts:250]: I think this is going to break in some nested directive situations. We need to try things like `guard([items], repeat(items, ...))`.

**Review by justinfagnani — CHANGES_REQUESTED** [review]

File: src/async-directive.ts
Comment (e111077) [reply src/async-directive.ts:30]: renaming to legacy. Peter said it used to be called disconnectable directive

File: src/async-directive.ts
Comment (e111077) [reply src/async-directive.ts:16]: this causes import name clashing and will have to still import `Part as LegacyPart`

File: src/directive.ts
Comment (e111077) [reply src/directive.ts:134]: Google style guide

File: src/directive.ts
Comment (e111077) [reply src/directive.ts:256]: Heya had a quick chat with peter and russell and we were a bit confused on what the concern is. Is it you think it'll write to the wrong location?

Additionally, we can't test with `guard` and `repeat` because they are not implemented using this class-based `Directive` impl. Additionally, assuming the user transferred their own directive to this and use it in one of our built-in ones, it should handle just fine.

The case where we think your concern about nested directives would come into play would be if a user implemented two of their own directives and nest them. I'd argue that's quite a fringe case.

I'd be happy to put a quick warning in a jsdoc saying this may not support certain advanced directive usage, but I'd like to understand the concern a bit more

File: src/directive-helpers.ts
Comment (bicknellr) [inline src/directive-helpers.ts:15]:
```suggestion
import {TemplateResult} from './lit-html.js';
export {isPrimitive} from './lib/parts.js';
```

Comment (bicknellr) [reply src/directive-helpers.ts:15]: (Adding this in http://cl/364917016)

File: src/directive.ts
Comment (kevinpschaaf) [reply src/directive.ts:256]: To clarify, I think:
* In 1.x, each type of directive hung its state off the part via weak map and it persisted even when directives changed and then changed back. Nested directives just worked by virtue of each having their own weak map on the part.
* In 2.x, each type of directive (by class) hangs its state off the part (or in the case of nesting, the nested directive), and its state is lost when directive type changes and then changes back.
* This PR currently matches the 2.x behavior as long as we don't consider nesting, but nesting will fail I think (but only for user directives, since the built-in directives don't use this, like Elliot said). Not sure if that's better or worse than caching by class type, which would make it more backward compatible (but doesn't get the clearing behavior of 2.x)

Comment (justinfagnani) [reply src/directive.ts:256]: It's not just nesting. This would fail:

```ts
html`${cond ? fooDirective() : barDirective()}`
```

because Foo and Bar directive classes would be mixed up.

Do something like, but with proper undefined handling:
```ts
const partToInstance = directiveCache.get(directiveClass);
const cached = partToInstance.get(part);
```

**Review by justinfagnani — CHANGES_REQUESTED** [review]
Once the cache issue is fixed and there are some tests this looks good.

File: src/directive.ts
Comment (e111077) [reply src/directive.ts:256]: Just checked. This does not fail because there is a weakmap per instance of `directive(...)` call rather than a persistent one. Added tests to show this

File: src/async-directive.ts
Comment (sorvell) [inline src/async-directive.ts:51]: Can this be a bit simpler, like this?

`return !!(node?.isConnected);`

File: src/directive.ts
Comment (sorvell) [inline src/directive.ts:148]:
```suggestion
 * Because event options are passed when adding listeners, we must take care
```

File: src/test/directives/migration-directives_test.ts
Comment (sorvell) [inline src/test/directives/migration-directives_test.ts:98]: Should add a test verifying that directive state is maintained across re-renders.

**Review by sorvell — CHANGES_REQUESTED** [review]

**Review by sorvell — APPROVED** [review]
LGTM

**Review by kevinpschaaf — DISMISSED** [review]

**Comment by e111077** [comment]
Ping @justinfagnani to update review

File: src/directive.ts
Comment (justinfagnani) [reply src/directive.ts:136]: We don't use Google internal style guide, and we don't use non-standard syntactic features of TypeScript outside of the type system. So no constructor parameter properties, namespaces or enums. Please be consistent with this codebase.

**Review by justinfagnani — CHANGES_REQUESTED** [review]

**Review by justinfagnani — APPROVED** [review]

**Comment by e111077** [comment]
ugh i had to format and I need an LGTM again

**Review by justinfagnani — APPROVED** [review]

**Pull request merged**
